src/vmm/src/vstate/vcpu/mod.rs

```diff
@@ -639,29 +639,152 @@ mod tests {
 
     lazy_static! {
         static ref RESULT: Mutex<Option<std::result::Result<VcpuExit<'static>, errno::Error>>> =
-            Mutex::new(Some(Ok(VcpuExit::InternalError)));
+            Mutex::new(None);
     }
 
     impl Vcpu {
         pub fn emulate(&self) -> std::result::Result<VcpuExit, errno::Error> {
-            RESULT.lock().unwrap().take().unwrap()
+            let mut g = RESULT.lock().unwrap();
+            if (&*g).is_some() {
+                return g.take().unwrap();
+            }
+            Err(errno::Error::new(libc::SIGILL))
         }
     }
 
+    struct DummyDevice;
+    impl devices::BusDevice for DummyDevice {}
+
     #[test]
     fn test_run_emulation() {
-        let (_vm, vcpu, _vm_mem) = setup_vcpu(0x1000);
+        let (_vm, mut vcpu, _vm_mem) = setup_vcpu(0x1000);
+
+        {
+            let mut g = RESULT.lock().unwrap();
+            *g = Some(Ok(VcpuExit::Hlt));
+        }
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(format!("{:?}", res), "Ok(Stopped)");
+
+        {
+            let mut g = RESULT.lock().unwrap();
+            *g = Some(Ok(VcpuExit::Shutdown));
+        }
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(format!("{:?}", res), "Ok(Stopped)");
```

Comment:
Comparing against verbose debug/display string is brittle. If/when display/debug is changed, all these tests will break.

You want to validate that `assert_eq!(res, Ok(VcpuExit::Shutdown))`, but can't do it directly because `PartialEq` is not implemented.

So you can still use the approach in this PR, but generate the display/debug bytes instead of hardcoding them:

```rust
assert_eq!(format!("{:?}", res), format!("{:?}", Ok(VcpuExit::Shutdown)));
```